src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1458 @@
+                return RD_KAFKA_RESP_ERR__INVALID_ARG;
+        }
+        if (check_oauthbearer_extension_value(token_value, errstr,
+                errstr_size) == -1)
+                return RD_KAFKA_RESP_ERR__INVALID_ARG;
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                if (check_oauthbearer_extension_key(extensions[i], errstr,
+                        errstr_size) == -1 ||
+                        check_oauthbearer_extension_value(extensions[i + 1],
+                                errstr, errstr_size) == -1)
+                                return RD_KAFKA_RESP_ERR__INVALID_ARG;
+        }
+        rd_kafka_wrlock(rk);
+        RD_IF_FREE(rk->rk_oauthbearer->md_principal_name, rd_free);
+        rk->rk_oauthbearer->md_principal_name = rd_strdup(md_principal_name);
+        RD_IF_FREE(rk->rk_oauthbearer->token_value, rd_free);
+        rk->rk_oauthbearer->token_value = rd_strdup(token_value);
+        rk->rk_oauthbearer->md_lifetime_ms = md_lifetime_ms;
+        // Schedule a refresh 80% through its remaining lifetime
+        rk->rk_oauthbearer->refresh_after_ms = now_wallclock_millis + 0.8 *
+                (md_lifetime_ms - now_wallclock_millis);
+        rd_list_clear(&rk->rk_oauthbearer->extensions);
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                rd_list_add(&rk->rk_oauthbearer->extensions,
+                        rd_strtup_new(extensions[i], extensions[i + 1]));
+        }
+        RD_IF_FREE(rk->rk_oauthbearer->errstr, rd_free);
+        rk->rk_oauthbearer->errstr = NULL;
+        rd_kafka_wrunlock(rk);
+        rd_kafka_dbg(rk, SECURITY, "BRKMAIN",
+                "Waking up waiting brokers after setting token");
+        rd_kafka_all_brokers_wakeup(rk, RD_KAFKA_BROKER_STATE_TRY_CONNECT);
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+}
+
+/**
+ * @brief SASL/OAUTHBEARER token refresh failure indicator.
+ *
+ * @param rk Client instance.
+ * @param errstr the mandatory SASL/OAUTHBEARER error message to set on the
+ *  client.
+ * 
+ * The SASL/OAUTHBEARER token refresh callback or event handler must cause
+ * this method to be invoked upon failure, via
+ * rd_kafka_oauthbearer_set_token_failure().
+ * 
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success, otherwise
+ *          RD_KAFKA_RESP_ERR__STATE if SASL/OAUTHBEARER is enabled but is not
+ *              configured to be the client's authentication mechanism.
+ * 
+ * @sa oauthbearer_set_token()
+ */
+rd_kafka_resp_err_t oauthbearer_set_token_failure(rd_kafka_t *rk,
+                const char *errstr) {
+        /* Check if SASL/OAUTHBEARER is the configured auth mechanism */
+        if (!rk->rk_oauthbearer)
+                return RD_KAFKA_RESP_ERR__STATE;
+
+        rd_assert(errstr);
+        rd_kafka_wrlock(rk);
+        RD_IF_FREE(rk->rk_oauthbearer->errstr, rd_free);
+        rk->rk_oauthbearer->errstr = strdup(errstr);
+        /* Leave any existing token because it may have some life left */
```

Comment:
This method is explicitly meant for documenting a token retrieval failure, and that failure could be a transient problem.  If we blank out the existing token then the client cannot make new connections to brokers, and I don't think a potentially transient token endpoint outage should lead to that state.

The issue of revoking existing tokens is an interesting one.  If we want that possibility then I think we should add a separate method for it.  That being said, though, it feels to me that the brokers can't trust clients to revoke tokens (assuming some token revocation infrastructure is in place, which is a big assumption), and whatever token revocation mechanism is available probably has to be operating at the brokers rather than at the clients.

My gut says leave it the way it is.  Thoughts?